mistralrs-core/src/kbnf.rs

```diff
@@ -0,0 +1,103 @@
+use ahash::AHashMap;
+use anyhow::Result;
+use candle_core::{DType, Device, Tensor};
+use kbnf::{
+    engine_like::{AcceptTokenError, MaskLogitsError},
+    AcceptTokenResult, Engine, EngineLike, Token, Vocabulary,
+};
+use tokenizers::Tokenizer;
+
+pub struct KbnfGrammar {
+    engine: Engine,
+    vocab_size: usize,
+}
+
+pub enum KbnfGrammarBias {
+    /// Token was accepted, it can be added to the sequence. No need for resampling.
+    Accepted,
+    /// Token was rejected. Resample with these new logits.
+    /// The token sampled with the bias can be added to the sequence.
+    Resample { new_logits: Tensor },
+    /// Generation was finished, the token can be added to the sequence,
+    /// but no more generation is necessary.
+    FinishedGeneration,
+}
+
+impl KbnfGrammar {
+    pub fn new(grammar: &str, tokenizer: &Tokenizer) -> Result<Self> {
+        let tokenizer_vocab = tokenizer.get_vocab(true);
```

Comment:
I think we still need to "unpreprocess" the vocabulary or the control ASCII characters(like `\n`) and all non-ASCII characters will not be recognized correctly.